src/perf_decode.ml

```diff
@@ -0,0 +1,669 @@
+open! Core
+open! Async
+open! Import
+
+let saturating_sub_i64 a b =
+  match Int64.(to_int (a - b)) with
+  | None -> Int.max_value
+  | Some offset -> offset
+;;
+
+let int64_of_hex_string str =
+  (* Bit hacks for fast parsing of hex strings.
+   *
+   * Note that in ASCII, ('1' | 'a' | 'A') & 0xF = 1.
+   *
+   * So for each character, take the bottom 4 bits, and add 9 if it's
+   * not a digit. *)
+  let res = ref 0L in
+  for i = 0 to String.length str - 1 do
+    let open Int64 in
+    let c = of_int (Char.to_int (String.unsafe_get str i)) in
+    res := (!res lsl 4) lor ((c land 0xFL) + ((c lsr 6) lor ((c lsr 3) land 0x8L)))
+  done;
+  !res
+;;
+
+let%test_module _ =
+  (module struct
+    open Core
+
+    let check str = Core.print_s ([%sexp_of: Int64.Hex.t] (int64_of_hex_string str))
+
+    let%expect_test "int64 hex parsing" =
+      check "fF";
+      [%expect {| 0xff |}];
+      check "f0f";
+      [%expect {| 0xf0f |}];
+      check "fA0f";
+      [%expect {| 0xfa0f |}];
+      check "0";
+      [%expect {| 0x0 |}]
+    ;;
+  end)
+;;
+
+let ok_perf_sample_line_re =
+  Re.Perl.re {|^ *([0-9]+)/([0-9]+) +([0-9]+).([0-9]+): *$|} |> Re.compile
+;;
+
+let ok_perf_sample_callstack_entry_re = Re.Perl.re "^\t *([0-9a-f]+) (.*)$" |> Re.compile
+
+let ok_perf_line_re =
+  Re.Perl.re
+    {|^ *([0-9]+)/([0-9]+) +([0-9]+).([0-9]+): +(call|return|tr strt|syscall|sysret|hw int|iret|tr end|tr strt tr end|tr end  (?:call|return|syscall|sysret|iret)|jmp|jcc) +([0-9a-f]+) (.*) => +([0-9a-f]+) (.*)$|}
+  |> Re.compile
+;;
+
+(* This matches exactly the power events which contain either [cbr] or [psb offs]. *)
+let ok_perf_power_line_re =
+  Re.Perl.re
+    {|^ *([0-9]+)/([0-9]+) +([0-9]+).([0-9]+): +([a-z]*)? +(cbr|psb offs): ([0-9]+ freq: ([0-9]+) MHz)?(.*)$|}
+  |> Re.compile
+;;
+
+let trace_error_re =
+  Re.Posix.re
+    {|^ instruction trace error type [0-9]+ (time ([0-9]+)\.([0-9]+) )?cpu [\-0-9]+ pid ([\-0-9]+) tid ([\-0-9]+) ip (0x[0-9a-fA-F]+|0) code [0-9]+: (.*)$|}
+  |> Re.compile
+;;
+
+let symbol_and_offset_re = Re.Perl.re {|^(.*)\+(0x[0-9a-f]+)\s+\(.*\)$|} |> Re.compile
+let unknown_symbol_dso_re = Re.Perl.re {|^\[unknown\]\s+\((.*)\)|} |> Re.compile
+
+type classification =
+  | Trace_error
+  | Ok_perf_line
+  | Ok_perf_sample_line
+  | Ok_perf_power_line
+
+let classify line =
+  if String.is_prefix line ~prefix:" instruction trace error"
+  then Re.Group.all (Re.exec trace_error_re line), Trace_error
+  else (
+    try Re.Group.all (Re.exec ok_perf_line_re line), Ok_perf_line with
+    | _ ->
+      (try Re.Group.all (Re.exec ok_perf_power_line_re line), Ok_perf_power_line with
+      | _ -> Re.Group.all (Re.exec ok_perf_sample_line_re line), Ok_perf_sample_line))
+;;
+
+let parse_time ~time_hi ~time_lo =
+  let time_lo =
+    (* In practice, [time_lo] seems to always be 9 decimal places, but it seems good
+         to guard against other possibilities. *)
```

Comment:
Nit: this module got de-indented so I think all the comments in this file have wonky spacing now.

Reply:
Should be fixed now